tutorial/README.md

```diff
@@ -0,0 +1,824 @@
+
+The database username, password and name should be the same as what we used for the `db` component parameters, but what about the hostname and port?
+
+- The port that we used for the `db` `Component` is `5432`
+- The hostname should be `db` - this will resolve via DNS due to the fact that we're running in Kubernetes
+
+Here's what those definitions might look like (your parameter names might be different):
+
+```js
+resource oam 'ApplicationConfiguration@core.oam.dev/v1alpha2' _ {
+  // .. metadata omitted
+  spec : {
+    components: [
+      {
+        componentName: 'db'
+        parameterValues: [
+          {
+            name: 'dbuser'
+            value: 'myuser'
+          }
+          {
+            name: 'dbpassword'
+            value: 'mypassword'
+          }
+          {
+            name: 'dbname'
+            value: 'hackfest'
+          }
+        ]
+      }
+      {
+        componentName: 'data-api'
+        parameterValues: [
+          {
+            name: 'dbuser'
+            value: 'myuser'
+          }
+          {
+            name: 'dbpassword'
+            value: 'mypassword'
+          }
+          {
+            name: 'dbname'
+            value: 'hackfest'
+          }
+          {
+            name: 'dbhostname'
+            value: 'db'
+          }
+          {
+            name: 'dbport'
+            value: '5432'
+          }
+        ]
+      }
+    ]
+  }
+}
+```
+
+This doesn't feel great, because we've got duplicated values. It would be easy to change one of these and not the other. We can introduce variables to hold these values so they have a single definition.
+
+Here's the same example using the `variable` statement to define variables:
+
+```js
+variable dbuser 'myuser'
+variable dbpassword 'mypassword'
+variable dbname 'hackfest'
+
+resource oam 'ApplicationConfiguration@core.oam.dev/v1alpha2' _ {
+  // .. metadata omitted
+  spec : {
+    components: [
+      {
+        componentName: 'db'
+        parameterValues: [
+          {
+            name: 'dbuser'
+            value: dbuser
+          }
+          {
+            name: 'dbpassword'
+            value: dbpassword
+          }
+          {
+            name: 'dbname'
+            value: dbname
+          }
+        ]
+      }
+      {
+        componentName: 'data-api'
+        parameterValues: [
+          {
+            name: 'dbuser'
+            value: dbuser
+          }
+          {
+            name: 'dbpassword'
+            value: dbpassword
+          }
+          {
+            name: 'dbname'
+            value: dbname
+          }
+          {
+            name: 'dbhostname'
+            value: 'db'
+          }
+          {
+            name: 'dbport'
+            value: '5432'
+          }
+        ]
+      }
+    ]
+  }
+}
+```
+
+Save these changes to the `ApplicationConfiguration` and we're ready to move on to the next step.
+
+### 4. Create the flights Component
```

Comment:
I like these exercises, mention at some point that they can use azoam by itself before applying, so they can catch syntax error upfront.

Reply:
Good idea, in fact **such** a good idea that I already did. This is explained in the first part that has you do `azoam` and again in the troubleshooting section.

Reply:
I mean to have it as a step, I did not see any step of running azoam by itself. The idea is to make it almost like spoon  feeding it.

Reply:
I don't want to include this as an explicit step because we're not going to explain the kubernetes output.

Reply:
So, there is an error message output from the command. So, we don't need this.